onnxscript/function_libs/torch_lib/ops/core.py

```diff
@@ -2157,12 +2157,54 @@ def aten_diagflat(self: TensorType, offset: int = 0) -> TensorType:
     raise NotImplementedError()
 
 
-def aten_diagonal(
-    self: TensorType, offset: int = 0, dim1: int = 0, dim2: int = 1
-) -> TensorType:
+@torch_op("aten::diagonal", trace_only=True)
+def aten_diagonal(self: TTensor, offset: int = 0, dim1: int = 0, dim2: int = 1) -> TTensor:
     """diagonal(Tensor(a) self, int offset=0, int dim1=0, int dim2=1) -> Tensor(a)"""
 
-    raise NotImplementedError()
+    self_rank = len(self.shape)
+    # If rank=2, then axes=[0]; if rank=3, then axes=[1]
+    axes = [self_rank - 2]
+    # [0,1,2] -> [2,0,1] when dim1=0 and dim2=1
```

Comment:
Looks like this can be scripted with https://github.com/onnx/onnx/issues/4981? I would add a todo if that's the case